### ISSUE TITLE
Closed Eyes, Open Mouths and Teeth

### ISSUE THREAD

**Issue by JayantPythonLover** [deepfakes/faceswap#296]
More of a discussion thread rather than an issue ... but has anyone looked into how we can improve the below mentioned features.

**Closed Eyes**
With the current models, if the source eyes are closed (generally during a frame where they blink), the target image is still rendered with open eyes. As a result all results look like the target is creepily staring at us or look absolutely dead and lifeless.

**Eye Detail**
Not just that but also the eye detail is pretty bad. In most cases they're just plain black or dark brown and that works but for sources with lighter eyes, the target eyes are one flat color and end up looking really spooky too.

**Closed Mouths**
Any source image with the mouth open beyond a certain point totally messes up the training process. Those images are more or less unusable and as a result any movement on the face that extends the jaw due to the mouth opening ends up looking stretched or weird because the target mouth is always almost closed or far less open.

**Teeth**
Teeth is often rendered as a series of white pixels with no black pixels in between to make for the divisions or give texture. This is not a problem generally when the subject is further away from the camera but for closeup shots, it ends up looking really bad.

--------

These are some of the glaring issues in terms of output that we face right now. I am thinking of working on some methods to improve these conditions.

Just wanted to know if any of you guys did any work in these aspects and if anyone had any idea that might be good to check out to address these issues and improve the model.

**Comment by Clorr** [comment]
I played a bit with loss function and res_blocks to get better details, however I had not much luck. In this try, however I got [better defined eyes](https://github.com/deepfakes/faceswap/issues/247#issuecomment-372017633)

**Comment by JayantPythonLover** [comment]
@Clorr That is actually great work. I can see the pupils inside the irises in most of them. Definitely a step in the right direction to addressing this issue.

I'll take a look at the model and some papers to see if there is anything we can do to address these issues. Fixing these will monumentally increase the quality of the output imo.

**Comment by Apollo122** [comment]
You probably know but perception loss helps with realistic eye movements -so it's said in gan repo anyways.

**Comment by Clorr** [comment]
In the crowd of super resolution nets, [this one](https://github.com/jormeli/srcnn-keras) could be very simple to test as it is implemented in keras.

**Comment by Apollo122** [comment]
You can check this too
https://github.com/titu1994/Image-Super-Resolution

**Comment by Clorr** [comment]
Also has anyone an idea about what PixelShuffler does? AFAIK it brings some noise which prevents overfitting, but do we know how much noise this is? Couldn't this perturbate the reconstruction as it is on each upscale layer?

**Comment by Clorr** [comment]
(Seems to be from SRGAN, but I couldn't find the theoric part around it. Some links pointed to tf.space_to_depth so maybe it does not introduce noise but it is more like a reshape. If you have more info I'd be glad to learn more about it)

**Comment by Kirin-kun** [comment]
We already have the pupils, from the source pictures. Since it's not something that has to be warped, wouldn't it be possible to recolor the visible portions using opencv?

Or train a new NN for the eyes :)

**Comment by bryanlyon** [comment]
@Kirin-kun Yes, I really like the idea of adding eye replacement using the
source image.  Pupils really don't need to change that much due to lighting
(we probably wont notice dilation).  I can work on this idea once my plate
is a little more clear.

**Comment by kvrooman** [comment]
Alternatively, fitting a 69th landmark for the pupil location and
performing the respective transforms and inverse transforms when
normalizing the faces would likely provide some improvement

**Comment by bryanlyon** [comment]
Yeah, that's about the idea I had. I would build a pupil recognition system and replace the original eye into the target eye.

**Comment by Kirin-kun** [comment]
pupil color can be affected by the lighting environment, like shadow over the face, especially for clear eyes, so a bulk replacement with a single reference pupil would probably produce sometimes unnatural pupils. 

So, ideally, you'd have to first analyze the lighting condition of the unmodified target eye of A and adapt the original eye from B face to it before replacing. Can be tricky if A and B have really different eye colors (dark vs clear). And that would probably be very expensive at convert time.

But personally, I don't mind long convert time if the result is more lifelike.

**Comment by Kirin-kun** [comment]
Actually, it's already been done for small areas, with the PatchMatch algorithm.

https://github.com/matthew-cox/PatchMatch-CUDA

Remove the pupil and the teeth and let a tool replace them according to a reference image like the one in the preview.

**Comment by kvrooman** [comment]
closing due to time. also the implementation of new models and more res blocks have ameliorated pupils and teeth to some extent.
